apps/passport-client/src/useSyncE2EEStorage.tsx

```diff
@@ -198,32 +244,90 @@ export async function downloadStorage(
     return { value: null, success: true };
   }
 
+  // Deserialize downloaded storage, which becomes the default new state if no
+  // merge is necessary.
+  const downloaded = await tryDeserializeNewStorage(
+    storageResult.value.storage
+  );
+  if (downloaded === undefined) {
+    return { error: null, success: false };
+  }
+  const { dlPCDs, dlSubscriptions, dlServerHash } = downloaded;
+
+  // Check if local app state has changes since the last server revision, in
+  // which case a merge is necessary.  Otherwise we keep the downloaded state.
+  let [newPCDs, newSubscriptions] = [dlPCDs, dlSubscriptions];
+  if (knownServerRevision !== undefined && knownServerHash !== undefined) {
+    const appStorage = await serializeStorage(
+      appSelf,
+      appPCDs,
+      appSubscriptions
+    );
+    if (appStorage.storageHash !== knownServerHash) {
+      const mergeResult = await mergeStorage(
+        { pcds: appPCDs, subscriptions: appSubscriptions },
+        {
+          pcds: downloaded.dlPCDs,
+          subscriptions: downloaded.dlSubscriptions
+        },
+        appSelf
+      );
+      if (!mergeResult.success) {
+        console.error(
+          "[SYNC] unable to merge new e2ee storage",
+          mergeResult.error
+        );
+        return { error: null, success: false };
+      }
+      ({ pcds: newPCDs, subscriptions: newSubscriptions } = mergeResult.value);
+    }
+  }
+
+  // We've successfully either accepted a new state, or created a merged state.
+  // Save and return results.
+  await savePCDs(newPCDs);
+  await saveSubscriptions(newSubscriptions);
+  savePersistentSyncStatus({
+    serverStorageRevision: storageResult.value.revision,
+    serverStorageHash: dlServerHash
+  });
+  console.log(
+    `[SYNC] downloaded e2ee storage (revision ${storageResult.value.revision})`
+  );
+  return {
+    value: {
+      pcds: newPCDs,
+      subscriptions: newSubscriptions,
+      serverRevision: storageResult.value.revision,
+      serverHash: dlServerHash
+    },
+    success: true
+  };
+}
+
+export async function tryDeserializeNewStorage(
+  storage: SyncedEncryptedStorage
+): Promise<
+  | undefined
+  | {
+      dlPCDs: PCDCollection;
+      dlSubscriptions: FeedSubscriptionManager;
+      dlServerHash: string;
+    }
+> {
```

Comment:
I wonder if using the type `APIResult` (which could be renamed to `MaybeResult`) could make sense here, given the way you're using it above? Another thing I wonder is if the type including fields starting with `dl` in the section I'm highlighting deserves a name, or if we could reuse an existing type.

Reply:
I don't think there's an existing type for this particular combination.  The inclusion of hash, or revision, or both is a little bit specific to each use case.

This function is really a single-use helper I introduced to make the flow of the calling function more clear and readable (due to the details of how the try/catch blocks needed to be set up).  Given that nature, I don't think it's worth extra weight to introduce just the right abstraction here.  In some other languages, I'd make it private or static, or prefix its name with an _ to clarify that.  Is there a pattern I could use here, or do you think we should adopt one?  I suppose it could be in a local closure, but that would make the function longer rather than shorter, and I already think it's too long.

Reply:
I think defining this function at the module scope is fine.